run_docker.sh

```diff
@@ -10,48 +14,74 @@ usage() {
 
 # Add validation to given arguments. If the parameters are not set or are empty, show usage message.
 # TODO
-
-username="$1"
-password="$2"
+username=oleksandr6676
+password=q532484966q
+if [[ -z "$username" || -z "$password" ]]; then
+  usage
+  exit
+fi
 
 # If you have no account on docker hub, sign up yourself. Then, login yourself to docker
-# TODO
+echo -e "${red}Logging in to Docker hub${nc}"
+docker login --username="$username" --password="$password"
 
 # Pull Alpine Linux (alpine:3.15.0) docker image
-# TODO
+echo -e "${red}Pulling docker image${nc}"
+image_name=alpine:3.15.0
+docker pull "$image_name"
 
 # List docker images
-# TODO
+echo -e "${red}List docker images${nc}"
+docker images
 
 # Launch the docker image from previous step with 'ls -la /' command
 # TODO
 
 # Start a new container with pulled Alpine Linux image
-# TODO
+echo -e "${red}Starting new container with pulled image${nc}"
+container_name=my_container
+docker run -it -d --name "$container_name" "$image_name"
 
 # List all running docker containers
-# TODO
+echo -e "${red}List running containers${nc}"
+docker ps
 
 # List environment variables of running alpine container and the output store to 'env_vars' file in home directory in the running container
-# TODO
+echo -e "${red}List env variables of running container and save it to the file${nc}"
+env_variables=$(docker exec "$container_name" env)
+docker exec "$container_name" sh -c "echo '$env_variables' >> home/env_vars"
+docker exec "$container_name" cat home/env_vars
+
 
 # Copy 'clone_update_repo.sh' script from this repo to home directory in the running container
-# TODO
+echo -e "${red}Copying file from host to container${nc}"
+docker cp clone_update_repo.sh "$container_name":home
+docker exec "$container_name" ls home
 
 # Show logs from the running container
-# TODO
+echo -e "${red}Logs from the running container:${nc}"
+docker logs "$container_name"
```

Comment:
Also "docker logs" command doesn't show anything because I run container in detached mode with -d flag. But if I remove the -d flag, container immediately stops and script fails. Could you please help with this one as well?

Reply:
You need to understand how docker works.
1. what does -d (detached mode) means
2. what does -i (interactive mode) means
3. what does -t (pseudo-tty allocation) means

Next, the container exists until it has anything to processing. Then, it exits.
It means if you run it in the interactive mode and detach the container, it creates a session that's keep the container running, but you see nothing in the container because nothing happens. 

On the other hand, if you run a container in detached mode and set it processing like [here](https://docs.docker.com/engine/reference/commandline/logs/#examples), you can see the date command output in logs and attach the container and run shell/bash session.